File: app/models/Notification.py
from app.instances.db import db
from app.notifications import get_title
from app.notifications import get_body
from app.helpers.SerializableEnum import SerializableEnum
from uuid import UUID
from M2Crypto.m2 import rand_bytes
import datetime

class NotificationType(SerializableEnum):
    STATUS_UPDATE = 0
    NEW_ANSWER = 1
    OUTGOLFED = 2
    NEW_POST_COMMENT = 3
    NEW_ANSWER_COMMENT = 4
    ANSWER_VOTE = 5
    POST_VOTE = 6

class NotificationStatus(SerializableEnum):
    UNSEEN = 0
    SEEN = 1
    READ = 2

class Notification(db.Model):
    """
    Represents a notification sent to a user
    """

    __tablename__ = 'notifications'
    __table_args__ = {'extend_existing': True}

    id = db.Column(db.String(36), primary_key=True, default=lambda: str(UUID(bytes=rand_bytes(16))))

    recipient_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    recipient = db.relationship('User', foreign_keys=[recipient_id], backref='notifications', lazy='joined')

    sender_id = db.Column(db.Integer, db.ForeignKey('users.id'), nullable=False)
    sender = db.relationship('User', foreign_keys=[sender_id], backref='sent_notifications', lazy='joined')

    # The type of notification see the NotificationType enum class
    notification_type = db.Column(db.Enum(NotificationType), nullable=False)

    # An ID referencing the item which dispatches the notification
    target_id = db.Column(db.Integer, nullable=True)

    # The id of the subscription source
    source_id = db.Column(db.Integer, nullable=True)

    date_created = db.Column(db.DateTime, default=datetime.datetime.now)

    read = db.Column(db.Enum(NotificationStatus), default=NotificationStatus.UNSEEN)


    def get_target_descriptor(self):
        """
        Returns the 'category' of the type's associated
        payload. This is used in the responder router
        """
        return {
            NotificationType.STATUS_UPDATE: 'status',
            NotificationType.NEW_ANSWER: 'answer',
            NotificationType.OUTGOLFED: 'answer',
            NotificationType.NEW_POST_COMMENT: 'post_comment',
            NotificationType.NEW_ANSWER_COMMENT: 'answer_comment',
            NotificationType.ANSWER_VOTE: 'answer',
            NotificationType.POST_VOTE: 'answer'
        }[self.notification_type]

    def get_title(self):
        # TODO: more descriptive titles
        return {
            NotificationType.STATUS_UPDATE: lambda: "Status Update",
            NotificationType.NEW_ANSWER: lambda: get_title.new_answer(self),
            NotificationType.OUTGOLFED: lambda: get_title.outgolfed(self),
            NotificationType.NEW_POST_COMMENT: lambda: get_title.post_comment(self),
            NotificationType.NEW_ANSWER_COMMENT: lambda: get_title.answer_comment(self),
            NotificationType.ANSWER_VOTE: lambda: "Someone voted on your answer",
            NotificationType.POST_VOTE: lambda: "Someone voted on your post"
        }[self.notification_type]()

    def get_body(self):
        # TODO: more descriptive bodies
        return {
            NotificationType.STATUS_UPDATE: lambda: "You're received a brand new status update.",
            NotificationType.NEW_ANSWER: lambda: get_body.new_answer(self),
            NotificationType.OUTGOLFED: lambda: get_body.outgolfed(self),
            NotificationType.NEW_POST_COMMENT: lambda: "A new comment has been posted on your challenge.",
            NotificationType.NEW_ANSWER_COMMENT: lambda: "A new comment has been posted on your answer.",
            NotificationType.ANSWER_VOTE: lambda: "A new vote has come upon your answer.",
            NotificationType.POST_VOTE: lambda: "A new vote has come upon your challenge."
        }[self.notification_type]()

    def get_plural(self):
        """
        Gets plural of notification type
        """
        return {
            NotificationType.STATUS_UPDATE: lambda: "updates",
            NotificationType.NEW_ANSWER: lambda: "new answers",
            NotificationType.OUTGOLFED: lambda: "outgolfs",
            NotificationType.NEW_POST_COMMENT: lambda: "new comments",
            NotificationType.NEW_ANSWER_COMMENT: lambda: "new comments",
            NotificationType.ANSWER_VOTE: lambda: "votes",
            NotificationType.POST_VOTE: lambda: "votes"
        }[self.notification_type]()

    def to_apns_json(self):
        """
        Returns APNS compliant JSON payload
        """
        target = str(self.target_id) if self.target_id is not None else '_'

        return {
            'aps': {
                'alert': {
                    'title': self.get_title(),
                    'body': self.get_body(),
                    'action': 'View'
                },
                'url-args': [self.id, self.get_target_descriptor(), target]
            }
        }

    def to_json(self):
        return {
            'id': self.id,
            'title': self.get_title(),
            'body': self.get_body(),
            'plural': self.get_plural(),
            'recipient': self.recipient.to_json(),
            'source_id': self.source_id,
            'sender': self.sender.to_json(),
            'target_id': self.target_id,
            'category': self.get_target_descriptor(),
            'date_created': self.date_created.isoformat(),
            'type': self.notification_type.value,
            'status': self.read.value
        }

    def __repr__(self):
        return "<Notification about {} for {}>".format(self.notification_type.name, self.recipient.name)

File: js/models/Notification.js
import RemoteEnum, { EnumEndpoint } from '~/models/Request/RemoteEnum';
import Theme from '~/models/Theme';
import User from '~/models/User';

/**
 * See {@link Notification.getTypes}
 * @typedef {number} NotificationType
 */

/**
 * See {@link Notification.getStatuses}
 * @typedef {number} NotificationStatus
 */

/**
 * @typedef {Object} NotificationCategory
 * @property {string} status
 * @property {string} answer
 * @property {string} postComment
 * @property {string} answerComment
 */
export const NotificationCategory = {
    status: 'status',
    answer: 'answer',
    postComment: 'post_comment',
    answerComment: 'answer_comment'
};

/**
 * Represents a notification
 * @implements {JSONConvertable}
 */
export default class Notification {

    /**
     * This returns an enumeration of notification types
     * @return {Enum} of the {@link NotificationType} enumeration
     */
    static async getTypes() {
        return await new RemoteEnum(EnumEndpoint.notificationType).run();
    }

    /**
     * This returns an enumeration of notification statuses
     * @return {Enum} of the {@link NotificationStatus} enumeration
     */
    static async getStatuses() {
        return await new RemoteEnum(EnumEndpoint.notificationStatus).run();
    }

    /**
     * A notification
     * @param {string} options.id - Notification UUID
     * @param {string} options.b
     * @param {string} options.body
     * @param {User} options.recipient - Delivery user
     * @param {User} options.sender - User who sent
     * @param {number} options.target - The ID of the target which triggered the notif
     * @param {number} options.source - the ID of the source which caused the context for recieving the notification
     * @param {string} options.category - The category based on a string for responder
     * @param {Date} options.dateCreated - The date created
     * @param {NotificationType} options.type - the notification type
     * @param {NotificationStatus} options.status - Read status
     */
    constructor({
        id,
        title,
        body,
        recipient,
        sender,
        target,
        source,
        category,
        dateCreated,
        type,
        status
    }) {
        this._id = id;
        this._title = title;
        this._body = body;
        this._recipient = recipient;
        this._sender = sender;
        this._target = target;
        this._category = category;
        this._source = source;
        this._dateCreated = dateCreated;
        this._type = type;
        this._status = status;
    }

    /**
     * @param {Object} json
     * @return {Notification}
     */
    static fromJSON(json) {
        return new Notification({
            id: json.id,
            title: json.title,
            body: json.body,
            recipient: User.fromJSON(json.recipient),
            sender: User.fromJSON(json.sender),
            target: json.target_id,
            source: json.source_id,
            category: json.category,
            dateCreated: new Date(json.date_created),
            type: json.type,
            status: json.status
        })
    }

    /**
     * Obtains responder URL
     * @readonly
     * @type {string}
     */
    get responder() {
        return `/responder/${this.id}/${this.category}/${this.target}`;
    }

    /**
     * Obtains Icon URL
     * @return {string}
     */
    async getIconURL() {
        const NotificationType = await Notification.getTypes();
        const key = NotificationType.keyForValue(this.type);
        return Theme.current.staticImageForTheme(`notification-icon/${key}`);
    }

    /**
     * Gets the body
     * @type {string}
     */
    get body() { return this._body; }

    /**
     * Gets the title
     * @type {string}
     */
    getTitle() { return this._title; }

    /**
     * The notification UUID
     * @readonly
     * @type {string}
     */
    get id() { return this._id; }

    /**
     * Obtains the category
     * @readonly
     * @type {NotificationCategory}
     */
    get category() { return this._category; }

    /**
     * The notification recipient
     * @readonly
     * @type {User}
     */
    get recipient() { return this._recipient; }

    /**
     * The notification sender
     * @readonly
     * @type {User}
     */
    get sender() { return this._sender; }

    /**
     * The target ID
     * @readonly
     * @type {number}
     */
    get target() { return this._target; }

    /**
     * The source ID. This is the post that resulted in the
     * notification subscription. For example in "A new answer
     * to your post". This would be the Post ID.
     * @readonly
     * @type {number}
     */
    get source() { return this._source; }

    /**
     * The notification creationdate
     * @readonly
     * @type {Date}
     */
    get dateCreated() { return this._dateCreated; }

    /*
     * The notification type
     * @readonly
     * @type {NotificationType}
     */
    get type() { return this._type; }

    /*
     * The notification read status
     * @type {NotificationStatus}
     */
    get status() { return this._status; }

    /**
     * Sets the status of this notification
     * @type {NotificationStatus}
     */
    set status(newStatus) { this._status = newStatus; }

    /**
     * Obtains the payload of the notification if possible.
     * @return {T} type depends on the notification
     */
    async getPayload() {
        // TODO:
    }

}
